deployments/apiv2/services/mantarray/src/main.py

```diff
@@ -124,55 +142,220 @@ async def delete_serial_number(
         try:
-            row = await con.fetchrow("SELECT hw_version FROM MAUnits WHERE serial_number=$1", serial_number)
-            hardware_version = row["hw_version"]
-            bind_threadlocal(hardware_version=hardware_version)
+            all_compatible_versions = await con.fetch(query, serial_number)
         except Exception:
-            err_msg = f"Serial Number {serial_number} not found"
-            logger.exception(err_msg)
-            return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": err_msg})
-    # try to get latest FW versions from HW version
+            logger.exception(f"Serial Number {serial_number} not found")
+            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST)
+
     try:
-        latest_versions = resolve_versions(hardware_version)
-        return JSONResponse({"latest_versions": latest_versions})
+        latest_versions = get_latest_compatible_versions(all_compatible_versions)
     except Exception:
-        err_msg = f"Error getting latest FW versions for {serial_number} with HW version {hardware_version}"
+        err_msg = f"Error determining latest FW + SW versions for {serial_number}"
         logger.exception(err_msg)
         return JSONResponse(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, content={"message": err_msg})
 
+    return latest_versions
+
+
+@app.get("/firmware/info")
+async def get_all_fw_sw_compatibility(
+    request: Request, token=Depends(ProtectedAny(scope=["mantarray:firmware:info"]))
+):
+    try:
+        async with request.state.pgpool.acquire() as con:
+            main_fw_info = await con.fetch("SELECT * FROM ma_main_firmware")
+            channel_fw_info = await con.fetch("SELECT * FROM ma_channel_firmware")
+
+        return FirmwareInfoResponse(main_fw_info=main_fw_info, channel_fw_info=channel_fw_info)
+    except Exception:
+        logger.exception("Error getting FW/SW compatibility")
+        raise
+
 
 @app.get("/firmware/{fw_type}/{version}")
 async def get_firmware_download_url(
-    fw_type: str = Path(..., regex="^(main|channel)$"),
-    version: str = Path(..., regex=r"^\d+\.\d+\.\d+$"),
+    fw_type: str = Path(..., regex=FW_TYPE_REGEX),
+    version: str = Path(..., regex=SEMVER_REGEX),
     token=Depends(ProtectedAny(scope=["mantarray:firmware:get"])),
 ):
     try:
         bind_threadlocal(fw_type=fw_type, fw_version=version)
 
-        url = get_download_url(version, fw_type)
+        url = get_fw_download_url(version, fw_type)
         return JSONResponse({"presigned_url": url})
     except Exception:
         err_msg = f"{fw_type.title()} Firmware v{version} not found"
         logger.exception(err_msg)
         return JSONResponse(status_code=status.HTTP_404_NOT_FOUND, content={"message": err_msg})
+
+
+@app.post("/firmware/{fw_type}/{version}")
+async def upload_firmware_file(
+    request: Request,
+    details: MainFirmwareUploadRequest | ChannelFirmwareUploadRequest,
```

Comment:
What's the difference again between main firmware and channel firmware?

Reply:
The main FW runs only on the main microcontroller of the instrument, the channel FW runs on each of the microcontrollers under each well